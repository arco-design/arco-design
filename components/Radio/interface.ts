import { CSSProperties, ChangeEvent, ReactNode, HTMLAttributes } from 'react';

/**
 * @title Radio
 */
export interface RadioProps<T = any>
  extends Omit<HTMLAttributes<HTMLLabelElement>, 'children' | 'className' | 'onChange'> {
  style?: CSSProperties;
  className?: string | string[];
  /**
   * @zh 是否禁用
   * @en Whether the radio is disabled
   */
  disabled?: boolean;
  /**
   * @zh 控件的 `value`
   * @en The value of radio
   */
  value?: T;
  /**
   * @zh 是否选中（受控模式）
   * @en Whether the radio is checked (Controlled)
   */
  checked?: boolean;
  /**
   * @zh 初始是否选中
   * @en Whether the radio is initially selected
   */
  defaultChecked?: boolean;
  /**
   * @zh 值变化的回调
   * @en Callback when radio status change
   */
  onChange?: (checked: boolean, event: ChangeEvent) => void;
  children?: ReactNode | ((value: { checked: boolean }) => ReactNode);
}

/**
 * @title Radio.Group
 */
export interface RadioGroupProps {
  style?: CSSProperties;
  className?: string | string[];
  disabled?: boolean;
  /**
   * @zh `Radio` 的 name
   * @en `Radio`'s name attr
   */
  name?: string;
  /**
   * @zh 单选的类型，是单选还是按钮
   * @en type of `Radio`
   * @defaultValue radio
   */
  type?: 'radio' | 'button';
  /**
   * @zh 方向
   * @en Arrangement direction
   * @defaultValue horizontal
   */
  direction?: 'vertical' | 'horizontal';
  /**
   * @zh 按钮类型的单选框尺寸（只在按钮类型下生效）
   * @en The size of radio button style(Only effective under `button` type)
   */
  size?: 'small' | 'default' | 'large' | 'mini';
  mode?: 'outline' | 'fill';
  /**
   * @zh 点击单选的回调
   * @en Callback when radio status change
   */
  onChange?: (value: any, event: ChangeEvent) => void;
  /**
   * @zh 默认选中的值
   * @en To set default value
   */
  defaultValue?: any;
  /**
   * @zh 选中的值（受控模式）
   * @en To set value
   */
  value?: any;
  /**
   * @zh 以数组配置的形式来设置单选组
   * @en Set children options
   */
  options?: (string | number | { label: ReactNode; value: any; disabled?: boolean })[];
}

export interface RadioGroupContextProps {
  type: 'radio' | 'button';
  value?: any;
  disabled?: boolean;
  group?: boolean;
  name?: RadioGroupProps['name'];
  onChangeValue?: (value: any, event: ChangeEvent) => void;
}
