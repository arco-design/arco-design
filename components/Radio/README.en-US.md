`````
Component / Data Entry

# Radio

In a set of related and mutually exclusive data, the user can only select one option.
`````

%%Content%%

## API

### Radio

|Property|Description|Type|DefaultValue|
|---|---|---|---|
|checked|Whether the radio is checked (Controlled)|boolean |`-`|
|defaultChecked|Whether the radio is initially selected|boolean |`-`|
|disabled|Whether the radio is disabled|boolean |`-`|
|className|Additional css class|string \| string[] |`-`|
|style|Additional style|CSSProperties |`-`|
|value|The value of radio|T |`-`|
|onChange|Callback when radio status change|(checked: boolean, event: ChangeEvent) => void |`-`|

### Radio.Group

|Property|Description|Type|DefaultValue|
|---|---|---|---|
|name|`Radio`'s name attr|string |`-`|
|direction|Arrangement direction|'vertical' \| 'horizontal' |`horizontal`|
|size|The size of radio button style(Only effective under `button` type)|'small' \| 'default' \| 'large' \| 'mini' |`-`|
|type|type of `Radio`|'radio' \| 'button' |`radio`|
|className|Additional css class|string \| string[] |`-`|
|defaultValue|To set default value|any |`-`|
|options|Set children options|(string \| number \| { label: ReactNode; value: any; disabled?: boolean })[] |`-`|
|style|Additional style|CSSProperties |`-`|
|value|To set value|any |`-`|
|onChange|Callback when radio status change|(value: any, event: ChangeEvent) => void |`-`|
