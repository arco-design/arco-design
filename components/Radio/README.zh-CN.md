`````
组件 / 数据输入

# 单选框 Radio

在一组相关且互斥数据中，用户仅能选择一个选项。
`````

%%Content%%

## API

### Radio

|参数名|描述|类型|默认值|
|---|---|---|---|
|checked|是否选中（受控模式）|boolean |`-`|
|defaultChecked|初始是否选中|boolean |`-`|
|disabled|是否禁用|boolean |`-`|
|className|节点类名|string \| string[] |`-`|
|style|节点样式|CSSProperties |`-`|
|value|控件的 `value`|T |`-`|
|onChange|值变化的回调|(checked: boolean, event: ChangeEvent) => void |`-`|

### Radio.Group

|参数名|描述|类型|默认值|
|---|---|---|---|
|name|`Radio` 的 name|string |`-`|
|direction|方向|'vertical' \| 'horizontal' |`horizontal`|
|size|按钮类型的单选框尺寸（只在按钮类型下生效）|'small' \| 'default' \| 'large' \| 'mini' |`-`|
|type|单选的类型，是单选还是按钮|'radio' \| 'button' |`radio`|
|className|节点类名|string \| string[] |`-`|
|defaultValue|默认选中的值|any |`-`|
|options|以数组配置的形式来设置单选组|(string \| number \| { label: ReactNode; value: any; disabled?: boolean })[] |`-`|
|style|节点样式|CSSProperties |`-`|
|value|选中的值（受控模式）|any |`-`|
|onChange|点击单选的回调|(value: any, event: ChangeEvent) => void |`-`|
