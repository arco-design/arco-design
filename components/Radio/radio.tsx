import React, { useContext, useRef } from 'react';
import cs from '../_util/classNames';
import Group, { RadioGroupContext } from './group';
import { ConfigContext } from '../ConfigProvider';
import omit from '../_util/omit';
import useMergeValue from '../_util/hooks/useMergeValue';
import IconHover from '../_class/icon-hover';
import { RadioProps } from './interface';
import useMergeProps from '../_util/hooks/useMergeProps';
import { isFunction } from '../_util/is';

function Radio(baseProps: RadioProps) {
  const inputRef = useRef<HTMLInputElement>(null);
  const { getPrefixCls, componentConfig, rtl } = useContext(ConfigContext);
  const props = useMergeProps<RadioProps>(baseProps, {}, componentConfig?.Radio);

  const context = useContext(RadioGroupContext);

  const prefixCls = getPrefixCls('radio');

  const mergeProps = { ...props };

  if (context.group) {
    mergeProps.checked = context.value === props.value;
    mergeProps.disabled = !!(context.disabled || props.disabled);
  }

  const { disabled, children, value, style, className, ...rest } = mergeProps;

  const [checked, setChecked] = useMergeValue(false, {
    value: mergeProps.checked,
    defaultValue: mergeProps.defaultChecked,
  });

  const classNames = cs(
    `${prefixCls}${context.type === 'button' ? '-button' : ''}`,
    {
      [`${prefixCls}-checked`]: checked,
      [`${prefixCls}-disabled`]: disabled,
      [`${prefixCls}-rtl`]: rtl,
    },
    className
  );

  const onChange = (event) => {
    const { onChange, value } = mergeProps;
    if (disabled) {
      return;
    }
    if (context.group) {
      context.onChangeValue && context.onChangeValue(value, event);
    } else if (!('checked' in props) && !checked) {
      setChecked(true);
    }
    !checked && onChange && onChange(true, event);
  };

  const onLabelClick = React.useCallback(
    (e) => {
      if (isFunction(props.children)) {
        // 避免children中含有表单元素造成label无法触发input的onchange的情况
        e.preventDefault();
        inputRef.current && inputRef.current.click();
      }
      rest.onClick && rest.onClick(e);
    },
    [props.children, rest.onClick]
  );

  return (
    <label
      {...omit(rest, ['checked', 'onChange'])}
      onClick={onLabelClick}
      style={style}
      className={classNames}
    >
      <input
        ref={inputRef}
        disabled={disabled}
        value={value || ''}
        type="radio"
        {...(context.name ? { name: context.name } : {})}
        checked={checked}
        onChange={(event) => {
          event.persist();
          onChange(event);
        }}
        onClick={(e) => {
          e.stopPropagation();
        }}
      />
      {isFunction(children) ? (
        children({ checked })
      ) : context.type === 'radio' ? (
        <>
          <IconHover
            prefix={prefixCls}
            className={`${prefixCls}-mask-wrapper`}
            disabled={checked || disabled}
          >
            <div className={`${prefixCls}-mask`} />
          </IconHover>
          {children && <span className={`${prefixCls}-text`}>{children}</span>}
        </>
      ) : (
        context.type === 'button' && <span className={`${prefixCls}-button-inner`}>{children}</span>
      )}
    </label>
  );
}

Radio.__BYTE_RADIO = true;

Radio.displayName = 'Radio';

Radio.Group = Group;

Radio.GroupContext = RadioGroupContext;

export default Radio;

export { RadioProps };
