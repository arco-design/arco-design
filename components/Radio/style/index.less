@import './token.less';
@import '../../style/mixin.less';

@radio-prefix-cls: ~'@{prefix}-radio';

.@{radio-prefix-cls},
.@{radio-prefix-cls}-button {
  > input[type='radio'] {
    opacity: 0;
    width: 0;
    height: 0;
    position: absolute;
    top: 0;
    left: 0;

    &:focus-visible + .@{prefix}-radio-icon-hover::before {
      background-color: @icon-hover-color-bg;
      opacity: 1;
    }

    &:focus-visible + .@{prefix}-radio-button-inner {
      box-shadow: inset 0 0 0 2px @radio-color-box-shadow;
      border-radius: inherit;
    }
  }
}

.icon-hover(@radio-prefix-cls, @radio-layout-height, @radio-size-mask-height);

.@{radio-prefix-cls} {
  position: relative;
  cursor: pointer;
  padding-left: (@radio-size-mask-height / 2) - (@radio-layout-height / 2);
  display: inline-block;
  font-size: @radio-font-text-size;
  line-height: unset;

  // radio circle
  &-text {
    color: @radio-color-text;
    margin-left: @radio-margin-text-left;
  }

  &-mask-wrapper {
    vertical-align: middle;
    top: -0.09em;
    position: relative;
    line-height: 1;
  }

  &-mask {
    display: block;
    border: @radio-border-width solid @radio-color-border;
    box-sizing: border-box;
    height: @radio-layout-height;
    width: @radio-layout-height;
    border-radius: @radio-border-radius;
    position: relative;
    line-height: @radio-layout-height;

    &::after {
      display: inline-block;
      box-sizing: border-box;
      position: absolute;
      content: '';
      border-radius: @radio-border-radius;
      background-color: @radio-color-bg;
      width: @radio-layout-height - @radio-border-width * 2;
      height: @radio-layout-height - @radio-border-width * 2;
      top: 0;
      left: 0;
      transform: scale(1);
      transition: transform @transition-duration-3 @transition-timing-function-overshoot;
    }
  }

  &:hover &-mask {
    border-color: @radio-color-border_hover;
  }

  &-checked &-mask {
    background-color: @radio-color-bg_checked;
    border-color: @radio-color-bg_checked;

    &::after {
      transform: scale(0.4);
      background-color: var(~'@{arco-cssvars-prefix}-color-white');
    }
  }

  &-checked:hover &-mask {
    border-color: @radio-color-bg_checked;
  }

  &-disabled {
    cursor: not-allowed;

    .@{prefix}-radio-icon-hover {
      cursor: not-allowed;
    }

    .@{radio-prefix-cls}-text {
      color: @radio-color-text_disabled;
    }
  }

  &-disabled &-mask {
    border-color: @radio-color-border_disabled;

    &::after {
      background-color: @radio-color-bg_disabled;
    }
  }

  &-disabled:hover &-mask {
    border-color: @radio-color-border_disabled;
  }

  &-checked&-disabled {
    .@{radio-prefix-cls}-mask,
    &:hover .@{radio-prefix-cls}-mask {
      border-color: @color-transparent;
      background-color: @radio-color-bg_checked_disabled;
    }

    .@{radio-prefix-cls}-mask::after {
      background-color: @radio-color-dot-bg_checked_disabled;
    }

    .@{radio-prefix-cls}-text {
      color: @radio-color-text_checked_disabled;
    }
  }

  &:hover {
    .@{prefix}-radio-icon-hover::before {
      background-color: @radio-mask-bg-color-bg;
    }
  }
}

// group type
.@{radio-prefix-cls}-group {
  display: inline-block;
  box-sizing: border-box;

  .@{radio-prefix-cls} {
    margin-right: @radio-group-margin-right;
  }

  // button type
  &-type-button {
    // ButtonGroup 中使用了 z-index: -1 的伪元素实现背景，为容器提供独立的层叠上下文以保证内部 Radio 表现正常
    position: relative;
    z-index: 0;
    padding: (@radio-button-spacing / 2);
    line-height: @radio-size-default-height - @radio-button-spacing * 2;
  }
}

.@{radio-prefix-cls}-button {
  display: inline-block;
  position: relative;
  margin: (@radio-button-spacing / 2);
  border-radius: @radio-button-border-radius;
  font-size: @radio-font-text-size;
  line-height: @radio-size-default-height - @radio-button-spacing * 2;
  color: @radio-button-color-text;
  background-color: @radio-button-color-bg;
  cursor: pointer;
  transition: all @transition-duration-1 @transition-timing-function-linear;

  &-inner {
    display: block;
    position: relative;
    padding: 0 @radio-button-padding-horizontal;
  }

  // 按钮之间的分割线
  &:not(:first-of-type)::before {
    position: absolute;
    top: 50%;
    left: -((@radio-button-spacing - @radio-button-size-separator-width) / 2) - @radio-button-size-separator-width;
    transform: translateY(-50%);
    display: block;
    height: @radio-button-size-separator-height;
    width: @radio-button-size-separator-width;
    background-color: @radio-button-color-separator-bg;
    content: '';
    transition: all @transition-duration-1 @transition-timing-function-linear;
  }

  // 被选中的按钮不需要左右的分割线
  &:hover::before,
  &:hover + &::before,
  &.@{radio-prefix-cls}-checked::before,
  &.@{radio-prefix-cls}-checked + &::before {
    opacity: 0;
  }

  // 按钮下层的背景色
  &::after {
    content: ' ';
    display: block;
    position: absolute;
    left: -@radio-button-spacing;
    top: $left;
    right: $left;
    bottom: $left;
    background-color: @radio-group-button-color-bg;
    pointer-events: none;
    z-index: -1;
  }

  &:first-of-type::after {
    border-top-left-radius: @radio-button-bg-border-radius;
    border-bottom-left-radius: @radio-button-bg-border-radius;
  }

  &:last-of-type::after {
    border-top-right-radius: @radio-button-bg-border-radius;
    border-bottom-right-radius: @radio-button-bg-border-radius;
  }

  &:hover {
    background-color: @radio-button-color-bg_hover;
    color: @radio-button-color-text_hover;
  }

  &.@{radio-prefix-cls}-checked {
    background-color: @radio-button-color-bg_active;
    color: @radio-button-color-text_active;
    font-weight: @radio-button-font-text-weight_active;
  }

  &.@{radio-prefix-cls}-disabled {
    cursor: not-allowed;
    background-color: @radio-button-color-bg_disabled;
    color: @radio-button-color-text_disabled;
  }

  &.@{radio-prefix-cls}-disabled.@{radio-prefix-cls}-checked {
    background-color: @radio-button-color-bg_checked_disabled;
    color: @radio-button-color-text_checked_disabled;
  }
}

// size small default large huge
.@{radio-prefix-cls} {
  &-size-small {
    line-height: @radio-size-small-height;

    &.@{radio-prefix-cls}-group-type-button,
    .@{radio-prefix-cls}-button {
      font-size: @radio-font-text-size_small;
      line-height: @radio-size-small-height - @radio-button-spacing * 2;
    }
  }

  &-size-large {
    line-height: @radio-size-large-height;

    &.@{radio-prefix-cls}-group-type-button,
    .@{radio-prefix-cls}-button {
      font-size: @radio-font-text-size_large;
      line-height: @radio-size-large-height - @radio-button-spacing * 2;
    }
  }

  &-size-mini {
    line-height: @radio-size-mini-height;

    &.@{radio-prefix-cls}-group-type-button,
    .@{radio-prefix-cls}-button {
      font-size: @radio-font-text-size_mini;
      line-height: @radio-size-mini-height - @radio-button-spacing * 2;
    }
  }
}

// vertical direction
.@{radio-prefix-cls}-group-direction-vertical {
  // default radio
  .@{radio-prefix-cls} {
    display: block;
    line-height: @radio-group-size-line-height_vertical;
    margin-right: 0;
  }
}

// dark mode

@{arco-theme-tag} {
  &[arco-theme='dark'] {
    .@{radio-prefix-cls}-button.@{radio-prefix-cls}-checked,
    .@{radio-prefix-cls}-button:not(.@{radio-prefix-cls}-disabled):hover {
      background-color: @radio-button-color-bg_active_dark;
    }

    .@{radio-prefix-cls}-button::after {
      background-color: @radio-group-button-color-bg_dark;
    }
  }
}

.@{radio-prefix-cls}-rtl {
  padding-left: 0;
  padding-right: (@radio-size-mask-height / 2) - (@radio-layout-height / 2);

  .@{radio-prefix-cls} {
    &-text {
      margin-left: 0;
      margin-right: @radio-margin-text-left;
    }
  }
}

.@{radio-prefix-cls}-group-rtl {
  .@{radio-prefix-cls} {
    margin-right: 0;
    margin-left: @radio-group-margin-right;
  }

  &.@{radio-prefix-cls}-group-direction-vertical {
    .@{radio-prefix-cls} {
      margin-right: unset;
      margin-left: 0;
    }
  }
}
