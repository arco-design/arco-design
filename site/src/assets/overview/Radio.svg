<svg width="128" height="128" viewBox="0 0 128 128" fill="none" xmlns="http://www.w3.org/2000/svg">
<rect x="53" y="53" width="22" height="22" rx="11" fill="#E8F3FF"/>
<rect x="60" y="60" width="8" height="8" rx="4" fill="#4080FF"/>
<rect x="53" y="53" width="22" height="22" rx="11" stroke="#4080FF" stroke-width="2"/>
</svg>
