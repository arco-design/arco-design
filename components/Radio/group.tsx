import React, { createContext, useContext, PropsWithChildren } from 'react';
import cs from '../_util/classNames';
import { isArray, isObject } from '../_util/is';
import Radio from './radio';
import { ConfigContext } from '../ConfigProvider';
import useMergeValue from '../_util/hooks/useMergeValue';
import { RadioGroupProps, RadioGroupContextProps } from './interface';
import useMergeProps from '../_util/hooks/useMergeProps';
import { pickDataAttributes } from '../_util/pick';
import { pickTriggerPropsFromRest } from '../_util/constant';

const defaultContextValue: RadioGroupContextProps = {
  type: 'radio',
};

const defaultProps: RadioGroupProps = {
  type: 'radio',
  mode: 'outline',
  direction: 'horizontal',
};

export const RadioGroupContext = createContext<RadioGroupContextProps>(defaultContextValue);

export const ClearRadioGroupContext = ({ children }: PropsWithChildren<{}>) => {
  return <RadioGroupContext.Provider children={children} value={defaultContextValue} />;
};

function Group(baseProps: PropsWithChildren<RadioGroupProps>) {
  const { getPrefixCls, size: ctxSize, componentConfig, rtl } = useContext(ConfigContext);
  const props = useMergeProps<PropsWithChildren<RadioGroupProps>>(
    baseProps,
    defaultProps,
    componentConfig?.['Radio.Group']
  );
  const { style, className, name, children, direction, type, mode, options, disabled } = props;

  const [value, setValue] = useMergeValue(undefined, {
    defaultValue: props.defaultValue,
    value: props.value,
  });
  const size = props.size || ctxSize;
  const prefixCls = getPrefixCls('radio');
  const classNames = cs(
    `${prefixCls}-group`,
    {
      [`${prefixCls}-group-type-button`]: type !== 'radio',
      [`${prefixCls}-size-${size}`]: !!size,
      [`${prefixCls}-mode-${mode}`]: !!mode,
      [`${prefixCls}-group-disabled`]: disabled,
      [`${prefixCls}-group-direction-vertical`]: direction === 'vertical',
      [`${prefixCls}-group-rtl`]: rtl,
    },
    className
  );

  const onChangeValue = (v: any, event): void => {
    const { onChange } = props;
    if (v !== value) {
      if (!('value' in props)) {
        setValue(v);
      }
      onChange && onChange(v, event);
    }
  };

  const contextProp = {
    onChangeValue,
    type,
    value,
    disabled,
    group: true,
    name,
  };
  return (
    <RadioGroupContext.Provider value={contextProp}>
      <div
        className={classNames}
        role="radiogroup"
        style={style}
        {...pickTriggerPropsFromRest(props)}
        {...pickDataAttributes(props)}
      >
        {options && isArray(options)
          ? options.map((option, index) => {
              if (isObject(option)) {
                return (
                  <Radio
                    key={option.value}
                    disabled={disabled || option.disabled}
                    value={option.value}
                  >
                    {option.label}
                  </Radio>
                );
              }
              return (
                <Radio key={index} value={option} disabled={disabled}>
                  {option}
                </Radio>
              );
            })
          : children}
      </div>
    </RadioGroupContext.Provider>
  );
}

Group.displayName = 'RadioGroup';

export default Group;

export { RadioGroupProps, RadioGroupContextProps };
