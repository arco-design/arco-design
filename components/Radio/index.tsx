import Radio, { RadioProps } from './radio';
import { RadioGroupProps } from './group';

export default Radio;

export { RadioProps, RadioGroupProps };

export { ClearRadioGroupContext } from './group';
