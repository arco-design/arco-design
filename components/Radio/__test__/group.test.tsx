import React from 'react';
import { mount } from 'enzyme';
import mountTest from '../../../tests/mountTest';
import componentConfigTest from '../../../tests/componentConfigTest';
import Radio from '..';

const RadioGroup = Radio.Group;

mountTest(RadioGroup);
componentConfigTest(RadioGroup, 'Radio.Group');

describe('Radio.Group', () => {
  function test(options, isChildren) {
    const children = isChildren
      ? options.map((x) => (
          <Radio key={x} value={x}>
            {x}
          </Radio>
        ))
      : null;
    it('expect render correctly', () => {
      const radioGroup = mount(
        <RadioGroup defaultValue="zhou" options={options}>
          {children}
        </RadioGroup>
      );
      expect(radioGroup.find('.arco-radio').length).toBe(3);

      expect(radioGroup.find('.arco-radio').at(0).hasClass('arco-radio-checked')).toBeTruthy();
    });
    it('expect onChange correctly', () => {
      const mockFn = jest.fn();
      const radioGroup = mount(
        <RadioGroup defaultValue="zhou" onChange={mockFn} options={options}>
          {children}
        </RadioGroup>
      );

      expect(radioGroup.find('.arco-radio').at(0).hasClass('arco-radio-checked')).toBeTruthy();

      radioGroup.find('.arco-radio input').at(2).simulate('change');

      expect(mockFn).toBeCalledTimes(1);
    });

    it('value control', () => {
      let value = 'zhou';
      const radioGroup = mount(
        <RadioGroup value={value} onChange={(v) => (value = v)} options={options}>
          {children}
        </RadioGroup>
      );
      expect(radioGroup.find('.arco-radio').at(0).hasClass('arco-radio-checked')).toBeTruthy();

      radioGroup.find('.arco-radio input').at(1).simulate('change');

      expect(value).toBe('jie');
      radioGroup.setProps({ value });

      expect(radioGroup.find('.arco-radio').at(1).hasClass('arco-radio-checked')).toBeTruthy();

      radioGroup.setProps({ value: 'lun' });

      expect(radioGroup.find('.arco-radio').at(2).hasClass('arco-radio-checked')).toBeTruthy();
    });
  }

  test(['zhou', 'jie', 'lun'], true);

  test(['zhou', 'jie', 'lun'], false);

  test(
    [
      { value: 'zhou', label: '周' },
      { value: 'jie', label: '杰' },
      { value: 'lun', label: '伦' },
    ],
    false
  );

  it('test disabled props', () => {
    const mockFn = jest.fn();
    const wrapper = mount(
      <RadioGroup
        onChange={mockFn}
        options={[
          { label: '1', value: '1' },
          { label: '2', value: '2' },
          { label: '3', value: '3', disabled: true },
        ]}
        defaultValue="2"
      />
    );

    expect(wrapper.find('.arco-radio').at(2).hasClass('arco-radio-disabled')).toBeTruthy();

    expect(wrapper.find('.arco-radio').at(1).hasClass('arco-radio-checked')).toBeTruthy();

    wrapper.find('.arco-radio input').at(0).simulate('change');
    expect(wrapper.find('.arco-radio').at(0).hasClass('arco-radio-checked')).toBeTruthy();
    expect(mockFn).toBeCalledTimes(1);

    wrapper.find('.arco-radio input').at(2).simulate('change');
    expect(wrapper.find('.arco-radio').at(0).hasClass('arco-radio-checked')).toBeTruthy();

    expect(mockFn).toBeCalledTimes(1);
  });
});
