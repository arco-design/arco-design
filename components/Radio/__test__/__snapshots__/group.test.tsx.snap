// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider componentConfig.Radio.Group default 1`] = `
<div
  class="arco-radio-group arco-radio-size-default arco-radio-mode-outline"
  role="radiogroup"
/>
`;

exports[`ConfigProvider componentConfig.Radio.Group set className globally 1`] = `
<div
  class="arco-radio-group arco-radio-size-default arco-radio-mode-outline global-default-class"
  role="radiogroup"
/>
`;

exports[`ConfigProvider componentConfig.Radio.Group set className globally and component set className itself. 1`] = `
<div
  class="arco-radio-group arco-radio-size-default arco-radio-mode-outline component-class"
  role="radiogroup"
/>
`;

exports[`ConfigProvider componentConfig.Radio.Group set data-* property 1`] = `
<div
  class="arco-radio-group arco-radio-size-default arco-radio-mode-outline"
  data-test-name="Radio.Group"
  role="radiogroup"
/>
`;
